### PR TITLE
Refactoring and added support for premium storage

### PR DESCRIPTION
**Pull request by trentmswanson**
* changed region to location
* removed scale grouping configurations and replaced with storage bin packing mapping
* included build templates and scripts for generating the data templates
* included the reference template test mocks
* added initial structure for simplifying the template usage

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@trentmswanson__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

File: elasticsearch/test/client-nodes-resources.json [hunk 1]
```diff
@@ -0,0 +1,101 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
```
Comment (singhkay) [inline elasticsearch/test/client-nodes-resources.json]: Is test an artifact left over from testing?

Comment (trentmswanson) [inline elasticsearch/test/client-nodes-resources.json]: No these are test resources when developing/testing some things in the top level azuredeploy.json template, they are essentially mocks that are generated from the referenced templates and deploy no resources.  The top level azuredeploy.json template can be changed to use the /test base path, and then you run a deployment you can test the parameter manipulation logic.  This makes editing the top level template much easier and faster.

Comment (trentmswanson) [inline elasticsearch/test/client-nodes-resources.json]: I guess after thinking about it here, I would have to say yes they are left over from test but intentionally included them.

**Pull request merged**